existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,264 @@
+
+```haskell
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+
+We might want to introduce some constraints.
+For example, we might want to specify that the `m` variable here is an instance of `Monad` typeclass.
+In order to add this constraint, we need to place it after all variables are declared:
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+```
+You might want to ask, what's so special about this syntax?
+Nothing changed since we added the quantifier because it was already there!
+However, there are a few things to notice here.
+
+### Examples
+
+#### Order of type variables
+
+First of all, you can now change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (yeah, there are real-life examples of such functions), and you know that when you use it, you want to specify the last argument explicitly.
+To pass a type as a parameter to a function, we will use the `TypeApplication` extension.
+Without `forall`, you would write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+Quite long, right?
+This is because GHC will maintain the same order as when these variables first appear in the function definition.
+However, by using explicit `forall.`, this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Support of other extensions
+
+`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall.` quantifier. :)
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = double ++ rest
+  where
+    double :: a -> [a]
+    double x = [x, x]
+```
+This will cause an error without `ScopedTypeVariables`. But to use it, you need to add a `forall.` quantifier.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how it was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+Besides universally quantified variables, Haskell also supports existential types.
+But in order to use them ... you still need to use the `forall.` keyword ¯\\\_(ツ)_/¯.
```

Comment:
I'm rooting for responsible journalism. [There was `exists`, but it was deprecated](https://web.archive.org/web/20060304191310/https://hackage.haskell.org/trac/haskell-prime/wiki/ExistentialQuantification#Syntaxofexistentials). Your post would better be, I guess, the only modern post about existentials to answer the question "why". :)

Reply:
Actually, Beside this article, I have seen only that it probably was added in jhc compiler, but that is it. No other clues, that it was actually added to ghc, or if it was removed from plans, without even implementing it